### PR TITLE
Update counter-cache on soft delete and undelete

### PR DESCRIPTION
**Pull request by kpandya91**

### CODE REVIEW

File: lib/soft_deletion/core.rb [hunk 1]
```diff
@@ -92,6 +92,13 @@ def soft_delete_dependencies
       self.class.soft_delete_dependents.map { |dependent| SoftDeletion::Dependency.new(self, dependent) }
     end
 
+    def update_counter_cache_associations(increment=false)
```
Comment (grosser) [inline lib/soft_deletion/core.rb]: prefer no default ... users should know what they want ... maybe make the argument `1 / -1`

Comment (grosser) [reply lib/soft_deletion/core.rb]: I guess 1/-1 won't really work ...

Comment (grosser) [inline lib/soft_deletion/core.rb]: should not be a public method ...
should be prefixed with soft_deletion or somehow have it in it's name ... 
also would prefer this as a single method ... logic is not so complicated it needs it's own class ...

Comment (kpandya91) [reply lib/soft_deletion/core.rb]: Yeah, `1 / -1` won't work so I just created an enumerator. Its easier to understand/read than passing in some value. How do you feel about that?

**Comment by kpandya91** [comment]
> Seems odd nobody found this bug sooner ... good catch :D

👍  :) Yeah, I was surprised it wasn't discussed in PR/issues before this.

The build fails for rails 3.2. Do we still support that version? If so, I can update the PR to get this to work for rails 3.2.

File: lib/soft_deletion/core.rb [hunk 2]
```diff
@@ -94,12 +94,26 @@ def soft_delete_dependencies
 
     protected
 
+    def soft_delete_counter_cache_associations
+      Enumerator.new do |yielder|
+        each_counter_cached_associations do |association|
+          foreign_key = association.reflection.foreign_key.to_sym
+          unless destroyed_by_association && destroyed_by_association.foreign_key.to_sym == foreign_key
+            if send(association.reflection.name)
+              yielder << association
```
Comment (grosser) [inline lib/soft_deletion/core.rb]: or just yield here and don't build an enumerator ? .. then `soft_delete_counter_cache_associations(&: decrement_counters)`

File: lib/soft_deletion/core.rb [hunk 3]
```diff
@@ -94,12 +94,24 @@ def soft_delete_dependencies
 
     protected
 
+    def soft_delete_counter_cache_associations
+      each_counter_cached_associations do |association|
+        foreign_key = association.reflection.foreign_key.to_sym
+        unless destroyed_by_association && destroyed_by_association.foreign_key.to_sym == foreign_key
```
Comment (grosser) [inline lib/soft_deletion/core.rb]: inlining the `foreign_key` would save the extra work to get the variable in the case it was not destroyed by an association ...

Comment (grosser) [reply lib/soft_deletion/core.rb]: like `destroyed_by_association.foreign_key.to_sym == association.reflection.foreign_key.to_sym`

File: lib/soft_deletion/core.rb [hunk 4]
```diff
@@ -94,12 +94,24 @@ def soft_delete_dependencies
...
+          if send(association.reflection.name)
```
Comment (grosser) [inline lib/soft_deletion/core.rb]: could this be simplified to `if association.target` or something like that ?

**Comment by grosser** [comment]
the `destroyed_by_association` code is kinda untested / it would work without, right ?
... I doubt it's actually necessary since a destroy would not cause a cascading soft_delete ...

**Comment by kpandya91** [comment]
Aah yes, it will work without the `destroyed_by_association` code since we don't override destroy here.

**Comment by grosser** [comment]
guess rails 3.2 did not like it ... maybe just opt-out ... or dig into AR 3.2 what it needs

**Comment by grosser** [comment]
1.2.0

**Comment by kpandya91** [comment]
Thanks @grosser!

**Pull request merged**